aten/src/ATen/TensorIterator.cpp

```diff
@@ -1281,14 +1294,51 @@ void TensorIteratorBase::build(TensorIteratorConfig& config) {
   view_offsets_ = DimVector(ndim_offsets, 0);
 }
 
+void TensorIteratorBase::set_output(int64_t output_idx, IntArrayRef sizes, IntArrayRef strides, TensorOptions options, DimnameList names) {
+  auto& op = operands_[output_idx];
+  TORCH_INTERNAL_ASSERT_DEBUG_ONLY(output_idx < num_outputs_);
+  const auto& t = maybe_get_output(output_idx);
+  TORCH_INTERNAL_ASSERT(t.defined());
+  if (!op.tensor.defined()) {
+    op.tensor = t;
+    op.current_dtype = op.target_dtype;
+  } else if (op.will_resize) {
+    // NB: op.tensor is NOT necessarily t in the else case; this
+    // can happen in the event of promotion, where t will correspond
+    // to the TRUE output, but op.tensor will be the different
+    // scalar type output
+    if (op.original_tensor.defined()) {
+      // Replay on op.tensor too!
```

Comment:
I don't really understand what's going on here - I'd gotten to the point of grokking the earlier, flatter version but I'm going to have to assume this logic is right

Reply:
This one is crushing and we might have to talk it through on VC. I was surprised when this one worked. It should be less complicated but I haven't figured out how to simplify it.